fde.md overhaul (#712)
void-linux/void-docs: src/installation/guides/fde.md

# Full Disk Encryption

> **Warning**: This guide uses `/dev/vda` as the example device name, Your drive's block device and other information may be different, so make sure it is correct.

Author: I changed sda to vda in a effort to make people pay closer attention to their drive's block device information, 
But i now think the better thing to do would be to change it to /dev/xxx1 or /dev/xxx2 instead of just droping it altogether.

> **Warning**: This guide uses `/dev/vda` as the example device name, Your drive's block device and other information may be different, so make sure it is correct.

Reviewer: We aren't here to save people from themselves when following a destructive guide that describes optional and advanced configuration. The example commands should be generally meaningful , not have bogus placeholders that require replacement.

## Partitioning

Boot a live image and login.

For a BIOS/MBR system, Create a single physical partition on the disk using
[cfdisk](https://man.voidlinux.org/cfdisk) with a `dos` label type, Mark it as bootable.  

> Now the partition layout should look similar to the following.

Author: Really, We are going to nit pick the wording change i chose?

I chose "similar to" instead of "like" because the readers partition layout may be different so as long as its **similar to** the example it should be fine.

Why drop BIOS/MBR? This guide doesnt cover BIOS/GPT so i thought the guide should specific.

> Now the partition layout should look similar to the following.

Reviewer: Yes, that's the point of a review. The point is to avoid making stylistic changes to an existing guide when there is no functional purpose for them.

BIOS/MBR is unnecessary because you already instruct people to make a 'dos' label, which is the nomenclature for MBR in the partition program.

partition. 
LVM is optional and may have its advantages or disadvantages depending on your system configuration.

The volume group name is not required to be the same as the LUKS partition name chosen above, 
But for consistency purposes it may be easier to use the same name for the volume group.

```
# vgcreate voidvm /dev/mapper/voidvm
```

There should now be an empty volume group named `voidvm`.

Next, logical volumes need to be created for the volume group. 
For this example, All of the available space is assigned to `/`.

The name chosen using `--name` should be noted because it will be used when mounting the partition, For example, `/dev/voidvm/root`.

```
# lvcreate --name root -l 100%FREE voidvm
```

### Filesystem creation

Next, create the filesystems. The example below uses Btrfs. Any filesystem [supported by
GRUB](https://www.gnu.org/software/grub/manual/grub/grub.html#Features) will
work.

On systems using LVM the root filesystem would be created on `/dev/voidvm/root`.

```
# mkfs.btrfs /dev/mapper/voidvm
```

On UEFI systems the EFI partition also needs the filesystem created.

```
# mkfs.vfat -F32 /dev/vda1

> ```

Author: 1. Im confused on this one, Are you proposing dropping LVM?

2. Following my guide as is i didnt really find this a problem, 
If you add Btrfs subvolumes this might be a problem, but i tested that and didnt see a problem, You should also remember that LVM is noted as optional and may have its advantages and disavantages, 

(2/3.) And Btrfs is noted as the example filesystem, nobody is forcing the user to use Btrfs

3. I changed the default filesystem to Btfs because it is the "modern" filesystem gaining widespread adoption, 
Using XFS makes this guide look really dated,

I think a solution here is to leave one as the example and note about a few other common filesystems 
(the most common would probably be Btrfs, ext4, F2FS and XFS, excluding ZFS due to it having a separate guide already).

## System installation

Next, mount the root partition, On a LVM configuration this would be `/dev/voidvm/root`

```
# mount /dev/mapper/voidvm /mnt
```

Next, some extra directories need to be mounted.

```
# for dir in dev proc sys run; do mkdir -p /mnt/$dir ; mount --rbind /$dir /mnt/$dir ; mount --make-rslave /mnt/$dir ; done
```

On a UEFI system, the `/efi` directory needs to be created and EFI system partition needs to be mounted.

```
# mkdir /mnt/efi

> # mount /dev/vda1 /mnt/efi

Reviewer: Drop all of these changes.

# cp /var/db/xbps/keys/* /mnt/var/db/xbps/keys/
```

Before entering the chroot to finish up configuration, it is time to perform the actual install.
Do not forget to use the [appropriate repository URL](../../xbps/repositories/index.md#the-main-repository) 
for the type of system you wish to install.

> Legacy BIOS systems use the `grub` package instead of `grub-x86_64-efi`. For LVM users, The `lvm2` package is required.

Reviewer: Drop this change.

```

> # xbps-install -Sy -R https://repo-default.voidlinux.org/current -r /mnt base-system cryptsetup grub-x86_64-efi

Reviewer: Drop this change.

```

Before entering the chroot, Copy a DNS configuration file from live system to allow internet connection in the chroot.

```

> # cp /etc/resolv.conf /mnt/etc/

Reviewer: Let's clean this up a bit:

For UEFI systems, also install necessary GRUB support.

```
xbps-install -Sy -R https://repo-default.voidlinux.org/current -r /mnt grub-x86_64-efi
```

```

Now you can enter the [chroot(1)](https://man.voidlinux.org/chroot.1) and continue configuring the system.

```
# chroot /mnt
# passwd root

> # echo AHOSTNAME > /etc/hostname

Reviewer: Drop this change, but add the resolv.conf copy before the chroot command.

The chown/chmod are probably superfluous, but it is almost costless to run them and potentially costly should the permissions be wrong.

Author: Did you read this part of the comment it opened the PR with:
'Changed example hostname from voidvm to AHOSTNAME to make it more apparent that the user should choose their own hostname.'

Your right, But i've run several installations in QEMU (UEFI and BIOS) and have yet to find a problem with not doing the chown/chmod.

### Filesystem configuration

Before continuing with filesystem configuration you need to get your [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier)s,
This can be done using utilities such as: [lsblk(8)](https://man.voidlinux.org/lsblk.8) or [blkid(8)](https://man.voidlinux.org/blkid.8).

The next step is editing `/etc/fstab`, 
The UUID needed for `/etc/fstab` is the `/dev/mapper/voidvm` UUID on non-LVM configurations and `/dev/voidvm/root` UUID on LVM configurations

```
# blkid -o value -s UUID /dev/mapper/voidvm
```

```
# <file system>	   <dir> <type>  <options>             <dump>  <pass>
tmpfs             /tmp  tmpfs   defaults,nosuid,nodev 0       0

> UUID=ADD-UUID-FROM-BLKID-COMMAND-HERE   /     btrfs   defaults              0       0

Reviewer: Drop these changes.

Author: Why? What is the point of dropping a change that tells users how to get their UUIDs?

Also the /etc/fstab should use the UUIDs not /dev/voidvm/root.

```

> 

Reviewer: Drop

UEFI systems will also have an entry for the EFI system partition.

```
# blkid -o value -s UUID /dev/vda1
```

```

> UUID=ADD-UUID-FROM-BLKID-COMMAND-HERE	/efi	vfat	defaults	0	0

Reviewer: Clean up a bit:

UEFI systems will also have an entry for the EFI system partition. This may look like:

```
/dev/sda1	/boot/efi	vfat	defaults	0	0
```

Alternatively, you can replace the device node `/dev/sda1` with a UUID, producing a line like:

```
UUID=<your-uuid>	/boot/efi	vfat	defaults	0	0
```

where the value of `<your-uuid>` is taken from the output of the command

```
# blkid -o value -s UUID /dev/sda1
```

Author: Not using the UUID can result in issues (#662).

```

### GRUB configuration

Next, configure GRUB to be able to unlock the encrypted filesystem:

```
# echo "GRUB_ENABLE_CRYPTODISK=y" >> /etc/default/grub
```

Next, the kernel needs to be configured to find the encrypted device. First,

> get the UUID of the device, Again this would be `/dev/vda1` on legacy BIOS systems.

Reviewer: ```suggestion
get the UUID of the device referred to by the `$LUKSPART` variable defined as described above.
```

```

> # blkid -o value -s UUID /dev/vda2

Reviewer: ```suggestion
# blkid -o value -s UUID "${LUKSPART}"
```

```

> Next, Edit the `GRUB_CMDLINE_LINUX_DEFAULT=` line in `/etc/default/grub` and add
> `rd.luks.uuid=ADD-UUID-FROM-BLKID-COMMAND-HERE` to it, If you are on a LVM configuration
> also add `rd.lvm.vg=voidvm`.

Reviewer: ```suggestion
Edit the `GRUB_CMDLINE_LINUX_DEFAULT=` line in `/etc/default/grub` and add
`rd.lvm.vg=voidvm rd.luks.uuid=<UUID>` to it. Make sure the UUID matches the one
for the [blkid(8)](https://man.voidlinux.org/blkid.8) command above.
```

# chmod -R g-rwx,o-rwx /boot
```

This keyfile also needs to be added to `/etc/crypttab`. Again, this would be
`/dev/vda1` on legacy BIOS systems.

```
# blkid -o value -s UUID /dev/vda2
```

```

> voidvm   UUID=ADD-UUID-FROM-BLKID-COMMAND-HERE   /boot/volume.key   luks

Reviewer: Refer to the `LUKSPART` variable to avoid repeatedly noting the distinction between BIOS and UEFI.

Next, install the boot loader to the disk.

For legacy BIOS systems:
```
# grub-install --target=i386-pc /dev/vda
```
For UEFI systems:
```

> # grub-install --target=x86_64-efi --efi-directory=/efi --bootloader-id="Void"

Reviewer: ```suggestion
# grub-install --target=x86_64-efi --bootloader-id="Void"
```
I assume it isn't necessary to specify a path to `/boot/efi` and I recommend against the change from `/boot/efi` to `/efi`. However, I don't know anything about GRUB, so `--efi-directory=/boot/efi` may still be required.

Author: This EFI directory is changed from /boot/efi to /efi to prevent a [Evil Maid](https://en.wikipedia.org/wiki/Evil_maid_attack) attack on the /boot partition (which would be encrypted if the EFI directory is /efi)

Reviewer: The disadvantage is that you have to specify the alternative path for GRUB.

The possibility of an Evil Maid attack has nothing to do with the mount point of the EFI system partition. You add no defense by mounting it somewhere else.

Reviewer 2: In this guide, `/boot` is encrypted already, so I'm not sure what you mean.  ESP is going to be subject to Evil Maid no matter where it gets mounted, since it'll never be encrypted.

# xbps-reconfigure -fa
```

Exit the `chroot`, unmount the filesystems. 

```
# exit
# umount -R /mnt

> ```
> 
> If you are using a LVM configuration you will need to run the following commands before continuing with the rest of the commands.
> 
> ```
> # lvchange -an voidvm
> 
> # cryptsetup close /dev/voidvm/root
> ```
> 
> Finally, Close the encrypted partition and reboot the system.
> 
> ```
> # cryptsetup close /dev/mapper/voidvm

Reviewer: Does any of this matter? If so:

```suggestion
# lvchange -an voidvm
# cryptsetup close /dev/mapper/voidvm
```

Otherwise, just drop all of this and go straight to the reboot.
